### DISCUSSION TITLE
Only allow specific fields in user profile to be updated

### DISCUSSION THREAD

**Discussion by theMackabu** [pocketbase/pocketbase#225] [Q&A] [answered]
Is it possible to allow specific fields to not be edited?
ive gotten about this far but cant seem to find a way to not allow those values to be changed

```go
	app.OnRecordBeforeUpdateRequest().Add(func(e * core.RecordUpdateEvent) error {
	    if (e.Record.TableName() == "profiles") {
	        fmt.Println(e.Record.GetStringDataValue("group"))
	        return nil
	    } else {
	        return nil
	    }
	})
```

this is also related to 'can some fields be hidden using the dao?'

```go

e.Router.AddRoute(echo.Route {
    Method: http.MethodGet,
    Path: "/api/profiles",
    Handler: func(c echo.Context) error {
        fieldResolver: = search.NewSimpleFieldResolver(
            "id", "created", "updated", "email", "verified",
        )

            users: = [] * models.User {}

            result,
        searchErr: = search.NewProvider(fieldResolver).
        Query(app.Dao().UserQuery()).
        ParseAndExec(c.QueryString(), & users)
        if searchErr != nil {
            return rest.NewBadRequestError("", searchErr)
        }

        if err: = app.Dao().LoadProfiles(users);err != nil {
            return rest.NewBadRequestError("", err)
        }

        event: = & core.UsersListEvent {
            HttpContext: c,
            Users: users,
            Result: result,
        }

            return app.OnUsersListRequest().Trigger(event, func(e * core.UsersListEvent) error {
            return e.HttpContext.JSON(http.StatusOK, e.Result)
        })
    },
    Middlewares: [] echo.MiddlewareFunc {
        apis.ActivityLogger(app),
    },
})

```

im so sorry Ive asked so many questions

**Comment by ganigeorgiev** [comment, marked as answer]
It depends on who you don't want to allow editing the field(s).

In general, there are 2 ways to prevent some fields to be updated:

1. With a collection update api rule:
    ```
    // where YOUR_FIELD is the name of the collection field that you don't want to allow updating
    @request.data.YOUR_FIELD = null || @request.data.YOUR_FIELD = YOUR_FIELD
    ```
2. With a hook:
    ```go
    app.OnRecordBeforeUpdateRequest().Add(func(e * core.RecordUpdateEvent) error {
        if (e.Record.Collection().Name != "profiles") {
            return nil
        }

        group = e.Record.GetStringDataValue("group");

        // compare against a fixed value
        if group != "YOUR_VALUE" {
            e.Record.SetDataValue("group", "YOUR_VALUE");
        }
        
        // OR if you want to compare to dynamic value, you can perform a query and feth the original record
        // (in future version I'll consider adding the original Record data to the event hook)
        originalRecord, err := app.Dao().FindRecordById(e.Record.Collection(), e.Record.Id, nil)
        if err != nil {
            return err
        }
        if group != originalRecord.GetStringDataValue("group") {
            // return an error or overwrite the request one
            e.Record.SetDataValue("group", originalRecord.GetStringDataValue("group"));
        }

        return nil
    })
    ```

About your second example - I'm not sure that I understand what you are trying to do. The endpoint `/api/profiles` from your example seems to be similar to the internal `/api/users` but without the access checks.

Is that what you want? Could you elaborate a little more on your use case?

**Reply by theMackabu** [reply]
I was trying to remove some keys from each user, such as lastVerifiedAt

**Reply by ganigeorgiev** [reply]
@theMackabu But why? With your custom endpoint `/api/profiles` everyone will be able to list your users and see their email address.

If you want to list some information of the user, my advice would be to add a `publicEmail` field in the user `profiles` collection and list only the records from the `profiles` collection (aka. not directly the user models). As a bonus, since the `profiles` collection is just like any other regular collection, you can define your access rules via the collection API rules settings without having to write a custom action.

**Reply by theMackabu** [reply]
thanks, it makes sense not to expose the private email

**Comment by jimafisk** [comment]
I have a scenario where I'd like field-level security (as opposed to collection-level security).

The requirements are:
- user profiles with `first`, `last`, `phone`, and `role` fields
- want new users to fill out their own profile besides `role`
- `role` is a multiple choice field with "member" and "staff" options, where "staff" will have elevated permissions on collections
- want to automatically set `role` to default value "member" when new user is creating their profile while ensuring they can't set it to "staff"
 
Would I create an API Rule on the profile collection like:

> Create Action: `userId = @request.user.id && role = "member"`

Then just make sure to send the default "member" value when running `client.Records.update()` in the custom app?

After profiles are filled out by users, I could manually set user profiles to "staff" in the Admin Dashboard. 
- once a user has a "staff" role, they should be able to update all user profiles in the custom app (including adding the "staff" role to other users)
- "member" users should be able to change their `first`, `last`, and `phone`, but not their `role`

Would that API Rule look like:

> Update Action: `(userId = @request.user.id && role = "member") || @request.data.role = "staff"`

Just want to make sure I'm understanding the rules correctly. Thank you!

**Reply by ganigeorgiev** [reply]
I think the create rule is OK, but the update rule will have to change because in your example the `staff` check has to be done for the current authenticated user, not based on the submitted request data:

```
(updated):
Wrong:   (userId = @request.user.id && role = "member") || @request.data.role = "staff"
Correct: (userId = @request.user.id && @request.data.role = "member") || @request.user.profile.role = "staff"
```

The first part allow only the owner of the profile record to modify all profile fields, except `role`.
The second part grant full access to users with `staff` role to change any field for any profile.

**Reply by jimafisk** [reply]
Ah yes, my mistake! Good catch, that makes sense to me. Thank you for the clarification!

**Reply by ganigeorgiev** [reply]
@jimafisk Actually, now that I think about it, the create rule for the profiles collection will never be used (unless it is called directly) because an empty profile will be created automatically in the background during the user creation. Hm, this probably will have to change...

**Reply by ganigeorgiev** [reply]
I'm not really sure what would be the best approach for the `profiles` create rule.
The empty profile is created primarily to simplify the internals and to reuse the existing Collection business logic.
Technically, since it is like any other collection, `profiles` create will be triggered for example 
if you call `client.records.delete("profiles", "RECORD_ID")` and then `client.records.create("profiles", { ... })`
but I can't think of a scenario where this would be useful...

From DX and UX prespective, maybe it will be better to add an exception for the `profiles` collection and hide/remove the create and delete rules.
I'll think more on this after the oauth2 changes for v0.6.0 release.

@jimafisk For your use case, the update rule will have to be further accommodated to take the empty profile in consideration:

```
(userId = @request.user.id && @request.data.role = "member") || @request.user.profile.role = "staff"
```

**Reply by jimafisk** [reply]
I was thinking about this too as I looked a little more closely, I was even referencing `client.Records.update()` in my "create" example not realizing I was actually running an "update." Thanks for the revised Update Action, I just gave it a quick test and it seemed to work! Is the difference being that `role` is the value that already exists in the database vs `@request.data.role` is the value you're submitting now?

Just thinking out loud (this might be a horrible suggestion, feel free to ignore), but when initially creating users, maybe you could optionally set profile information:
```js
let profile = {
  first: "",
  last: "",
  phone: "",
  role: "member"
};

const createUserAndProfile = async () => {
  try {
    await client.Users.create({
      email: email,
      password: password,
      passwordConfirm: passwordConfirm,
      profile: profile,
    });
  } catch (e) {
    console.log(e);
  }
}
```

That could kick off the Create Action for profiles, not sure if that's clear enough DX to what is actually happening. If you could do it in a way that the `profile` key is not required, it hopefully wouldn't break backwards compatibility. Honestly you're probably fine just disabling the unused Create Action if that's the path of least resistance. Thanks again for all the help!

**Reply by ganigeorgiev** [reply]
>  Is the difference being that role is the value that already exists in the database vs @request.data.role is the value you're submitting now?

Yes, `role` (or any other direct record field) for the `list`, `view`, `update` and `delete` actions holds the already persisted value. `@request.data.*` are the user submitted request body data.

For `create` - since we don't have a previous record to get the value from, the create rule actually is applied **after a [dummy transaction insert](https://github.com/pocketbase/pocketbase/blob/master/apis/record.go#L178-L204)** (without files upload and delete). If it fails, the transaction is rollbacked and and the record is not persisted in the database.

Thanks for the create user suggestion. I actually had something very similar in mind, but I dropped the idea because it is ambiguous which `@request.user` will be used when resolving the `profiles` create rule - the just created user or the already authenticated user (if `client.users.authViaEmail()` was called before that).

**Reply by MeharGaur** [reply]
My solution was to use hooks:
```go
        // Private field hooks
	app.Server.OnRecordViewRequest().Add(privatizeViewResponse)
	app.Server.OnRecordsListRequest().Add(privatizeListResponse)

	// Protected field hooks
	app.Server.OnRecordBeforeUpdateRequest().Add(protectProfileUpdate)
```

So I have a JSON field on the profile model called `private_fields`. That JSON array holds a list of field names that should not be leaked, and are only visible to the owner or an admin.

In the `privatizeResponse` functions I take the model data, strip out any fields that are in `private_fields`, and then after it has been cleaned I send my own HTTP JSON response and stop hook propagation, so that pocketbase doesn't send the same data twice.

In the protected fields, I just have a struct that holds which fields are not allowed to be updated by the user, like say a view count. If the request tries to modify a protected field, then ignore the request by stopping hook propagation.

That seems to be working so far for me, let me know if you see any issues or security flaws with that approach.

**Reply by ganigeorgiev** [reply]
@MeharGaur I'd personally go with a `relation` to a new collection that holds the hidden fields, but your solution using the hooks also seems good to me. There are several things that you need to be aware though:
- The `update` action also returns a Record response similar to the `view`, so you may need to strip your `private_fields` from there too (if you haven't already).
- The user view action can be accessed by the owner of the user model and since we load the related profile together with the user, you may want to also strip the `private_fields` with the `OnUserViewRequest` hook (_in v0.6 or v0.7 I will consider adding a check for the `profiles` view rule before embedding it inside the user model for consistency with the `expand` query parameter_).
- On auth response (email/oauth2 authentication, token refresh, confirm password reset, etc.) also returns the user model with embedded profile, so depending on your requirements whether the owner is allowed to view the `private_fields`, you may also have to attach to the `OnUserAuthRequest` hook and "clean" the `user.profile` from the hook data.

_btw. Because the models we send to the hooks are pointers, you can also directly modify the hook data in place in your `privatizeViewResponse` and `privatizeListResponse` and return nil without stopping the hook propagation (this is useful for the cases when your have multiple hook listeners)._

**Reply by ganigeorgiev** [reply]
@jimafisk, @MeharGaur, @theMackabu I've filed a proposal to merge the users and profiles into a single structure to cover some of discussed shortcomings - https://github.com/pocketbase/pocketbase/issues/376.

_If you find time, please review it and leave your feedback in case you have a different use case that may not be covered by the proposed changes or have better idea in general how to handle the users management._
